src/common/logging.cpp

```diff
@@ -28,64 +28,270 @@
 
 #include "logging.hpp"
 
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
 #include <assert.h>
 #include <errno.h>
 #include <stdarg.h>
 #include <stdint.h>
 #include <string.h>
 #include <syslog.h>
+#include <sys/time.h>
 
 static int        sLevel = LOG_INFO;
 static const char HEX_CHARS[] = "0123456789abcdef";
 
+static unsigned msecs_start;
+static bool     log_col0 = true; /* we start at col0 */
+static FILE    *log_fp;
+static bool     syslog_enabled = true;
+static bool     syslog_opened = false;
+
+#define LOGFLAG_syslog 1
+#define LOGFLAG_file   2
+
+/** Set/Clear syslog enable flag */
+void otbrLogEnableSyslog(bool b)
+{
+    syslog_enabled = b;
+}
+
+/** Enable logging to a specific file */
+void otbrLogSetFilename(const char *filename)
+{
+    log_fp = fopen(filename, "w");
+    if (log_fp == NULL)
+    {
+        fprintf(stderr, "Cannot open log file: %s\n", filename);
+        perror(filename);
+        exit(EXIT_FAILURE);
+    }
+}
+
+/** Get the current debug log level */
+int  otbrLogGetLevel(void)
+{
+    return sLevel;
+}
+
+/** Set the debug log level */
+void otbrLogSetLevel(int aLevel)
+{
+    assert(aLevel >= LOG_EMERG && aLevel <= LOG_DEBUG);
+    sLevel = aLevel;
+}
+
+/** Determine if we should not or not log, and if so where to */
+static int do_log(int aLevel)
+{
+    int r;
+
+    assert(aLevel >= LOG_EMERG && aLevel <= LOG_DEBUG);
+
+    r = 0;
+
+    if (syslog_opened && syslog_enabled && (aLevel <= sLevel))
+    {
+        r = r | LOGFLAG_syslog;
+    }
+
+    /* if someone has turned on the seperate file the most likely
+     * situation is they are debugging a problem, or need a extra
+     * information. In this case, we do not test the log level.
+     */
+    if (log_fp != NULL)
+    {
+        r = r | LOGFLAG_file;
+    }
+    return r;
+}
+
+
+/** return the time, in milliseconds since application start */
+static unsigned msecs_now(void)
```

Comment:
The only example I find like this - of a static function is:
```
static void MbedtlsDebug(void *aContext, int aLevel, const char *aFile, int aLine, const char *aMessage)
```
In the src/agent/dtls_mbed.cpp file

So I followed that style,